provider/akamai.go

```diff
@@ -0,0 +1,413 @@
+	NewRequest(config edgegrid.Config, method, path string, body io.Reader) (*http.Request, error)
+	Do(config edgegrid.Config, req *http.Request) (*http.Response, error)
+}
+
+type akamaiOpenClient struct{}
+
+func (*akamaiOpenClient) NewRequest(config edgegrid.Config, method, path string, body io.Reader) (*http.Request, error) {
+	return c.NewRequest(config, method, path, body)
+}
+
+func (*akamaiOpenClient) Do(config edgegrid.Config, req *http.Request) (*http.Response, error) {
+	return c.Do(config, req)
+}
+
+// AkamaiConfig clarifies the method signature
+type AkamaiConfig struct {
+	DomainFilter          DomainFilter
+	ZoneIDFilter          ZoneIDFilter
+	ServiceConsumerDomain string
+	ClientToken           string
+	ClientSecret          string
+	AccessToken           string
+	DryRun                bool
+}
+
+// AkamaiProvider implements the DNS provider for Akamai.
+type AkamaiProvider struct {
+	domainFilter DomainFilter
+	zoneIDFilter ZoneIDFilter
+	config       edgegrid.Config
+	dryRun       bool
+	client       akamaiClient
+}
+
+type akamaiError struct {
+	Title     string `json:"title"`
+	Status    int    `json:"status"`
+	Detail    string `json:"detail"`
+	RequestID string `json:"requestId"`
+}
+
+type akamaiZones struct {
+	Zones []akamaiZone `json:"zones"`
+}
+
+type akamaiZone struct {
+	ContractID string `json:"contractId"`
+	Zone       string `json:"zone"`
+}
+
+type akamaiRecordsets struct {
+	Recordsets []akamaiRecord `json:"recordsets"`
+}
+
+type akamaiRecord struct {
+	Name  string        `json:"name"`
+	Type  string        `json:"type"`
+	TTL   int64         `json:"ttl"`
+	Rdata []interface{} `json:"rdata"`
+}
+
+// NewAkamaiProvider initializes a new Akamai DNS based Provider.
+func NewAkamaiProvider(akamaiConfig AkamaiConfig) *AkamaiProvider {
+	edgeGridConfig := edgegrid.Config{
+		Host:         akamaiConfig.ServiceConsumerDomain,
+		ClientToken:  akamaiConfig.ClientToken,
+		ClientSecret: akamaiConfig.ClientSecret,
+		AccessToken:  akamaiConfig.AccessToken,
+		MaxBody:      1024,
+		HeaderToSign: []string{
+			"X-External-DNS",
+		},
+		Debug: false,
+	}
+
+	provider := &AkamaiProvider{
+		domainFilter: akamaiConfig.DomainFilter,
+		zoneIDFilter: akamaiConfig.ZoneIDFilter,
+		config:       edgeGridConfig,
+		dryRun:       akamaiConfig.DryRun,
+		client:       &akamaiOpenClient{},
+	}
+	return provider
+}
+
+func (p *AkamaiProvider) request(method, path string, body io.Reader) ([]byte, error) {
+	req, err := p.client.NewRequest(p.config, method, fmt.Sprintf("https://%s/%s", p.config.Host, path), body)
+	if err != nil {
+		log.Errorf("Akamai client failed to prepare the request")
+		return nil, err
+	}
+	resp, err := p.client.Do(p.config, req)
+	if err != nil {
+		log.Errorf("Akamai client failed to do the request")
+		return nil, err
+	}
+
+	//204 means no body on success by choice
+	if resp.StatusCode == 204 {
+		return nil, nil
+	}
+
+	defer resp.Body.Close()
+	byt, err := ioutil.ReadAll(resp.Body)
```

Comment:
I thought about it but decided not to. The 'request' function is only a wrapper for the Akamai client because of it's bad error handling. The variable byt is returned if no errors have happened and is then unmarshaled in other functions. So I don't know the struct yet. Maybe the byt gets used  fetchZones and it's zones struct or by fetchRecordSet and it's recordSet struct.

I thought about implementing the wrapper in a different way so that I could use json.NewDecoder but after reading:
https://stackoverflow.com/questions/21197239/decoding-json-using-json-unmarshal-vs-json-newdecoder-decode/21198571#21198571

I decided that it is not worth it, since I have to store the response right away because it is returned by the 'Do' function.
What do you think about this?

Reply:
ah, now i see what you mean. I even do an unmarshal in each following function. I will change it to a json decoder. I do know the struct each time.

Reply:
Okay, it is much cleaner code now. Thanks for this suggestion :+1: